gems/aws-sdk-s3/lib/aws-sdk-s3/file_downloader.rb

```diff
@@ -151,50 +128,61 @@ def multithreaded_get_by_parts(n_parts, total_size, etag)
 
       def download_in_threads(pending, total_size)
         threads = []
+        mutex = Mutex.new
+        max_requests = pending.count
+        total_requests = 0
         progress = MultipartProgress.new(pending, total_size, @progress_callback) if @progress_callback
         @thread_count.times do
           thread = Thread.new do
             begin
-              while part = pending.shift
+              while (part = pending.shift)
                 if progress
                   part.params[:on_chunk_received] =
                     proc do |_chunk, bytes, total|
                       progress.call(part.part_number, bytes, total)
                     end
                 end
                 resp = @client.get_object(part.params)
+
+                if part.params[:range]
+                  range = resp.content_range.split(' ').last.split('/').first
+                  expected_range = part.params[:range].split('=').last
+                  raise FileDownloadError, 'file download integrity checked failed' unless expected_range == range
+                end
```

Comment:
What is this being parsed and what is compared? Perhaps pulling this out into a method can better document what is happening.